src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -263,24 +268,32 @@ static void uct_rdmacm_cm_handle_event_established(struct rdma_cm_event *event)
     uct_rdmacm_cm_ep_t *cep = event->id->context;
 
     ucs_assert(event->id == cep->id);
-    cep->flags |= UCT_RDMACM_CM_EP_GOT_CONNECT;
-    uct_cm_ep_server_connect_cb(&cep->super, UCS_OK);
+    /* Do not notify user on disconnected EP, RDMACM out of order case */
+    if (cep->flags & UCT_RDMACM_CM_EP_GOT_DISCONNECT) {
+        return;
+    }
+
+    uct_rdmacm_cm_ep_server_connect_cb(cep, UCS_OK);
 }
 
 static void uct_rdmacm_cm_handle_event_disconnected(struct rdma_cm_event *event)
 {
-    uct_rdmacm_cm_ep_t *cep      = event->id->context;
-    struct sockaddr    *remote_addr = rdma_get_peer_addr(event->id);
-    char               ip_port_str[UCS_SOCKADDR_STRING_LEN];
-    char               ep_str[UCT_RDMACM_EP_STRING_LEN];
+    uct_rdmacm_cm_ep_t   *cep         = event->id->context;
+    struct sockaddr      *remote_addr = rdma_get_peer_addr(event->id);
+    char                 ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    char                 ep_str[UCT_RDMACM_EP_STRING_LEN];
+    uct_cm_remote_data_t remote_data;
 
     ucs_debug("%s: got disconnect event, status %d peer %s",
               uct_rdmacm_cm_ep_str(cep, ep_str, UCT_RDMACM_EP_STRING_LEN),
               event->status, ucs_sockaddr_str(remote_addr, ip_port_str,
                                              UCS_SOCKADDR_STRING_LEN));
 
     cep->flags |= UCT_RDMACM_CM_EP_GOT_DISCONNECT;
-    cep->super.disconnect_cb(&cep->super.super.super, cep->super.user_data);
+    /* calling error_cb instead of disconnect CB directly handles out-of-order
+     * disconnect event prior connect_response/connect_established event */
+    remote_data.field_mask = 0;
+    uct_rdmacm_cm_ep_error_cb(cep, &remote_data, UCS_ERR_CONNECTION_RESET);
```

Comment:
why need to change this to common function, just makes it more complicated, not the common function also has to check if status != RESET

Reply:
fixed thru https://github.com/openucx/ucx/pull/4719#discussion_r374078962